lib/segment/src/index/hnsw_index/graph_layers_builder.rs

```diff
@@ -82,85 +86,110 @@ impl GraphLayersBuilder {
-
+    pub fn subgraph_connectivity(&self, points: &[PointOffsetType], q: f32) -> f32 {
         if points.is_empty() {
             return 1.0;
         }
 
         let max_point_id = *points.iter().max().unwrap();
 
         let mut visited: BitVec = BitVec::repeat(false, max_point_id as usize + 1);
-        let mut bitmask: BitVec = BitVec::repeat(false, max_point_id as usize + 1);
+        let mut point_selection: BitVec = BitVec::repeat(false, max_point_id as usize + 1);
 
         for point_id in points {
-            bitmask.set(*point_id as usize, true);
+            point_selection.set(*point_id as usize, true);
         }
 
         let mut rnd = rand::rng();
 
-        let mut queue = Vec::new();
-
         // Try to get entry point from the entry points list
         // If not found, select the point with the highest level
         let entry_point = self
             .entry_points
             .lock()
             .get_random_entry_point(&mut rnd, |point_id| {
-                bitmask.get_bit(point_id as usize).unwrap_or(false)
+                point_selection.get_bit(point_id as usize).unwrap_or(false)
             })
             .map(|ep| ep.point_id);
 
         // Select entry point by selecting the point with the highest level
-
-        let entry_point = if let Some(entry_point) = entry_point {
-            // Entry point is found
-            entry_point
-        } else {
+        let entry_point = entry_point.unwrap_or_else(|| {
             points
                 .iter()
                 .max_by_key(|point_id| self.links_layers[**point_id as usize].len())
                 .cloned()
                 .unwrap()
-        };
+        });
+        let entry_layer = self.get_point_level(entry_point);
 
-        // Start BFS from the entry point
+        let mut queue: Vec<u32> = vec![];
 
-        queue.push(entry_point);
+        // Amount of points reached when searching the graph.
+        let mut reached_points = 1;
 
-        let already_visited = visited.replace(entry_point as usize, true);
-        if !already_visited {
-            reached_points += 1;
-        }
+        // Total points visited (also across retries).
+        let mut visited_points_total = 0;
 
-        // Do not skip edges on the first point links
-        // to avoid random noise
-        let mut first = true;
+        // Retry loop, in case some budget is left.
+        loop {
+            // Points visited in the previous layer (Get used as entry point in the iteration over the next layer)
+            let mut previous_visited_points = vec![entry_point];
 
-        while let Some(current_point) = queue.pop() {
-            let links = self.links_layers[current_point as usize][layer].read();
+            // For each layer in HNSW...
+            for current_layer in (0..=entry_layer).rev() {
+                // Set entry points to visited points of previous layer.
+                if previous_visited_points.is_empty() {
```

Comment:
let's refactor to make sure `previous_visited_points` actually never shrinks, so we always use point as entry if we ever visited it on previous layers